File: Python/net_protocols/high/http/har.py
from . import http
import base64

def from_req(obj):
	obj = obj['request']
	hdr = {}
	for h in obj['headers']:
		k = h['name']
		if k[0] == ':':
			continue
		http._hdr_add_(hdr, k.lower(), h['value'])
	inf = {}
	inf['type'] = 0
	inf['method'] = obj['method'].upper()
	url = obj['url']
	url = url[url.find("//")+2:]
	p = url.find("/")
	if not 'host' in hdr:
		hdr['host'] = url[:p]
	inf['url'] = url[p:]
	body = b''
	if 'postData' in obj:
		body = bytes(obj['postData']['text'],'utf8')
	sz = len(body)
	if sz != 0 or inf['method'] == 'POST':
		hdr['content-length'] = str(sz)
	return http.http(inf, "HTTP/1.1", hdr, body)
def from_resp(obj):
	obj = obj['response']
	code = obj['status']
	if code == 0:
		return None
	hdr = {}
	for h in obj['headers']:
		k = h['name'].lower()
		if (k == 'status'
		or k == 'content-encoding'
		or k == 'transfer-encoding'):
			continue
		http._hdr_add_(hdr, k, h['value'])
	inf = {}
	inf['type'] = 1
	inf['code'] = code
	inf['code_text'] = obj['statusText']
	body = b''
	if 'content' in obj:
		o = obj['content']
		if 'text' in o:
			text = o['text']
			if 'encoding' in o and o['encoding'] == 'base64':
				body = base64.b64decode(text)
			else:
				body = bytes(text,'utf8')
	sz = len(body)
	if sz != 0:
		hdr['content-length'] = str(sz)
	return http.http(inf, "HTTP/1.1", hdr, body)

File: Go/fls/parsefile/fl_har/har.go
package fl_har

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"io/ioutil"
	"net/http"
	"os"
	"strconv"
	"strings"
)

type header struct {
	Name  string
	Value string
}

type request struct {
	Method   string
	Url      string
	Headers  []header
	PostData struct {
		Text string
	}
}

type response struct {
	Status     int
	StatusText string
	Headers    []header
	Content    struct {
		Text     string
		Encoding string
	}
}

type entrie struct {
	Request  request
	Response response
}

func (e entrie) getReq() *http.Request {
	rq := e.Request
	req, _ := http.NewRequest(rq.Method, rq.Url, strings.NewReader(rq.PostData.Text))
	for _, hdr := range rq.Headers {
		name := hdr.Name
		if name[0] == ':' {
			continue
		}
		req.Header.Set(strings.Title(name), hdr.Value)
	}
	return req
}

func (e entrie) getResp() *http.Response {
	rs := e.Response
	if rs.Status == 0 {
		return nil
	}
	res := &http.Response{Proto: "HTTP/1.1", ProtoMajor: 1, ProtoMinor: 1, Header: make(http.Header)}
	res.StatusCode = rs.Status
	res.Status = rs.StatusText

	body := rs.Content.Text
	if body != "" {
		var data []byte
		if rs.Content.Encoding == "base64" {
			data, _ = base64.StdEncoding.DecodeString(body)
		} else {
			data = []byte(body)
		}
		res.ContentLength = int64(len(data))
		res.Body = ioutil.NopCloser(bytes.NewReader(data))
	}

	for _, hdr := range rs.Headers {
		name := strings.ToLower(hdr.Name)
		if name == "status" ||
			name == "content-encoding" ||
			name == "transfer-encoding" {
			continue
		}
		res.Header.Set(strings.Title(name), hdr.Value)
	}
	res.Header.Set("Content-Length", strconv.FormatInt(res.ContentLength, 10))
	return res
}

type har struct {
	Log struct {
		Entries []entrie
	}
}

func Parse(fname string) []http.Request {
	fl, err := os.Open(fname)
	if err != nil {
		return nil
	}
	var data har
	err = json.NewDecoder(fl).Decode(&data)
	if err != nil {
		return nil
	}
	var res []http.Request
	for _, entrie := range data.Log.Entries {
		req := entrie.getReq()
		req.Response = entrie.getResp()
		res = append(res, *req)
	}
	return res
}
